File: c/rna-transcription/rna_transcription.c
#include "rna_transcription.h"

#include <stdlib.h>
#include <string.h>

static char translate_nucleotide(char nucleotide)
{
    switch (nucleotide) {
        case 'G':
            return 'C';
        case 'C':
            return 'G';
        case 'T':
            return 'A';
        case 'A':
            return 'U';
        default:
            return '\0';
    }
}

char *to_rna(const char *dna)
{
    char *rna = (char *) malloc(strlen(dna) + 1);

    char *nucleotide = rna;
    for (; *dna != '\0'; ++dna) {
        *nucleotide++ = translate_nucleotide(*dna);
    }
    *nucleotide = '\0';

    return rna;
}

File: r/rna-transcription/rna-transcription.R
transcriptions <- c(G = "C", C = "G", T = "A", A = "U")
transcribe <- function(nucleotide) {
  output <- unname(transcriptions[nucleotide])
  stopifnot(!is.na(output))
  output
}

to_rna <- function(dna) {
  paste0(sapply(unlist(strsplit(dna, "")), transcribe), collapse = "")
}
